lib/authz/service.go

```diff
@@ -0,0 +1,79 @@
+package authz
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/authz/opa_files"
+	"github.com/enfabrica/enkit/lib/authz/plugins"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+)
+
+type regoInput struct {
+	Username string   `json:"username"`
+	Resource Resource `json:"resource"`
+	Action   Action   `json:"action"`
+}
+
+func (i regoInput) JSON() string {
+	js, _ := json.Marshal(i)
+	return string(js)
+}
+
+type AuthService struct {
+	q rego.PreparedEvalQuery
+}
+
+var (
+	ErrorOnlyAccessSelfOwned = errors.New("permission denied to perform this action")
+)
+
+func (as *AuthService) Do(ctx context.Context, req *AuthRequest) error {
+	i := &regoInput{
+		Username: req.UserName,
+		Resource: req.Resource,
+		Action:   req.Action,
+	}
+	res, err := as.q.Eval(ctx, rego.EvalInput(i))
+	if err != nil {
+		return err
+	}
+	fmt.Println(res)
+	if len(res) == 0 {
+		return ErrorOnlyAccessSelfOwned
```

Comment:
pretty cryptic? If it wasn't for the "permission denied to perform this action" string I would not know what this was. Better name for the error?